pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -0,0 +1,246 @@
+// indicate that the client has stopped expecting a response.
+const ProxyForwardTimeout = 10 * time.Second
+
+// DNSProxy is a L7 proxy for DNS traffic. It keeps a list of allowed DNS
+// lookups that can be regexps and blocks lookups that are not allowed.
+// A singleton is always running inside cilium-agent.
+type DNSProxy struct {
+	lock.Mutex
+
+	// BindAddr is the local address we bind to to listen for DNS requests.
+	// Note: unlike the other proxies, this server listens on a fixed, shared,
+	// port.
+	BindAddr string
+
+	// NotifyOnDNSResponse is the callback by which the proxy can emit DNS
+	// response data.
+	// Note: this is a little pointless since this proxy is in-process but it is
+	// intended to allow us to switch to an external proxy process by forcing the
+	// design now.
+	NotifyOnDNSResponse NotifyOnDNSResponseFunc
+
+	// UDPServer, TCPServer are the miekg/dns server instances. They handle DNS
+	// parsing etc. for us.
+	UDPServer, TCPServer *dns.Server
+
+	// UDPClient, TCPClient are the miekg/dns client instances. Forwarded
+	// requests are made with these clients but are sent to the originally
+	// intended DNS server.
+	// Note: The DNS request ID is randomized but when seeing a lot of traffic we
+	// may still exhaust the 16-bit ID space for our (source IP, source Port) and
+	// this may cause DNS disruption. A client pool may be better.
+	UDPClient, TCPClient *dns.Client
+
+	// allowed tracks all allowed matchNames. These are regexps, even simple
+	// matchNames with no regexp wildcards are compiled. We ensure a unique value
+	// per source policy of a matchName because the RegexpMap handles the
+	// reference counting of unique values but will de-dupe repeats.
+	// Note: Simple DNS names, e.g. bar.foo.com, will treat the "." as a literal.
+	// We convert these "." in regexps that only contain dots (and alphanumeric
+	// characters) into a regexp "." literal. This is because the more common use
+	// case will be these simple literals.
+	// To insert a wildcard ".", use .{1} to indicate a single wildcard character.
+	allowed *regexpmap.RegexpMap
+}
+
+type NotifyOnDNSResponseFunc func(lookupTime time.Time, name string, ips []net.IP, ttl int) error
+
+// StartDNSProxy starts a proxy used for DNS L7 redirects
+func StartDNSProxy(address string, port uint16, notifyFunc NotifyOnDNSResponseFunc) (*DNSProxy, error) {
+	if port == 0 {
+		return nil, errors.New("DNS proxy port not configured")
+	}
+
+	p := &DNSProxy{
+		BindAddr:            fmt.Sprintf("%s:%d", address, port),
+		NotifyOnDNSResponse: notifyFunc,
+		allowed:             regexpmap.NewRegexpMap(),
+	}
+
+	// Start the DNS listeners on UDP and TCP
+	p.UDPServer = &dns.Server{Addr: p.BindAddr, Net: "udp", Handler: p}
+	p.TCPServer = &dns.Server{Addr: p.BindAddr, Net: "tcp", Handler: p}
+
+	for _, s := range []*dns.Server{p.UDPServer, p.TCPServer} {
+		go func(server *dns.Server) {
+			if err := server.ListenAndServe(); err != nil {
```

Comment:
This will fail silently and will make it hard to debug when this fails. Can we retry and Fatal if we can't bind to the port after 5 tries over 20 seconds?

Reply:
Done! I think there is a `ActivateAndServer` where we setup the actual `Listener`. In that case we would do the bind before running the goroutine and be able to return the error in the constructor. I'll look into it but for now it does the fatal.